examples/paillier_logistic_regression.py

```diff
@@ -0,0 +1,245 @@
+                tar.write(remotedata.read())
+
+        with tarfile.open(foldertar) as tar:
+            tar.extractall('examples/')
+        os.remove(foldertar)
+
+
+def preprocess_data():
+    """
+    Get the Enron emails from disk.
+    Represent them as bag-of-words.
+    Shuffle and split train/test.
+    """
+
+    print("Importing dataset from disk...")
+    path = 'examples/enron1/ham/'
+    ham1 = [open(path + f, 'r', errors='replace').read().strip(r"\n")
+            for f in os.listdir(path) if os.path.isfile(path + f)]
+    path = 'examples/enron1/spam/'
+    spam1 = [open(path + f, 'r', errors='replace').read().strip(r"\n")
+             for f in os.listdir(path) if os.path.isfile(path + f)]
+    path = 'examples/enron2/ham/'
+    ham2 = [open(path + f, 'r', errors='replace').read().strip(r"\n")
+            for f in os.listdir(path) if os.path.isfile(path + f)]
+    path = 'examples/enron2/spam/'
+    spam2 = [open(path + f, 'r', errors='replace').read().strip(r"\n")
+             for f in os.listdir(path) if os.path.isfile(path + f)]
+
+    # Merge and create labels
+    X = ham1 + spam1 + ham2 + spam2
+    y = np.array([-1] * len(ham1) + [1] * len(spam1) +
+                 [-1] * len(ham2) + [1] * len(spam2))
+
+    # Words count, keep only fequent words
+    count_vect = CountVectorizer(decode_error='replace', stop_words='english',
+                                 min_df=0.001)
+    X = count_vect.fit_transform(X)
+
+    print('Vocabulary size: %d' % X.shape[1])
+
+    # Shuffle
+    perm = np.random.permutation(X.shape[0])
+    X, y = X[perm, :], y[perm]
+
+    # Split train and test
+    split = 500
+    X_train, X_test = X[-split:, :], X[:-split, :]
+    y_train, y_test = y[-split:], y[:-split]
+
+    print("Labels in trainset are %.2f spam / %.2f ham"
+          % (np.mean(y_train == 1), np.mean(y_train == -1)))
+
+    return X_train, y_train, X_test, y_test
+
+
+class TimeIt():
+    """Helper for measuring runtime"""
+
+    def tick(self):
+        self.time0 = time.time()
+
+    def tock(self):
+        if not self.time0:
+            raise Exception('Need to `tick` first!')
+
+        time1 = time.time() - self.time0
+        print('[elapsed time: %.2f s]' % time1)
+
+
+class PaillierClassifier():
+    """Scorer of emails with an encrypted models"""
+
+    def __init__(self, pubkey):
+        self.pubkey = pubkey
+
+    def set_weights(self, weights, intercept):
+        self.weights = np.array(weights)
+        self.intercept = intercept
+
+    def encrypted_predict(self, x):
+
+        score = self.intercept
+        _, idx = x.nonzero()
+        for i in idx:
+            score += x[0, i] * self.weights[i]
+        return score
+
+    def encrypted_evaluate(self, X):
+        return [self.encrypted_predict(X[i, :]) for i in np.arange(X.shape[0])]
+
+
+class Alice():
+    """
+    Train a model on clear data.
+    Possess the private key and can encrypt the model for remote usage.
```

Comment:
The more important thing about being the private key holder is Alice can decrypt encrypted results (or models).